### ISSUE TITLE
A happy user wishes to improve usability

### ISSUE THREAD

**Issue by HagaiHargil** [AllenInstitute/deepinterpolation#14]
Hi,

Our lab (Dr. Pablo Blinder, Tel Aviv University) is very impressed from the results of this model. Analysis of 2P calcium imaging data before and after the noise reduction increases the number of detected active components quite dramatically, and happily enough all of them seem completely valid.

The issues we do have with the package are on the user-friendliness side of things - documentation, high-level API, etc. Are there any plans to work on that? Of course I'll be more than happy to help improving things once their prioritized.

**Comment by jeromelecoq** [comment]
I am glad to hear things are working with our models on your ends! Several groups have duplicated our results so it is really encouraging!

Regarding code usability. There are many ways to improve things. I was not entirely sure going into this publication what would be better for dissemination : Cloud deployment, Github repo, Local clusters, Model servers, ...  I tried to release as much as I could to see what gains more traction.  I have been thinking of opening up a poll or such to better understand the pain points.

**Comment by HagaiHargil** [comment]
Thanks for the response.

To be clearer, I was referring to the actual API of the package, rather than the deployment side of things. Indeed there are many routes to take when you wish to deploy it - and it seems that you covered a whole bunch of them! -  but then you still have to run the code itself and "hit" the usability issues I referred to.

To be even more specific, I believe that detailed walkthroughs are needed, which will be based on the example scripts and notebooks but will contain much more information on the purpose of each parameter and possible fine-tunings that can be done. Then you might want some high-level API that can run the model in a step or two, given the current configuration which can be set somewhere else. This high-level function can receive some `Config` object or file that contains the needed parameters. This can make the package easier to work with when you're trying to automate things, or to plug it in into some computational pipeline.

There are many ways of doing these things - sadly each ML\DL framework works slightly differently - but I think that there's a case for adopting at least _one_ of these methods. I'll be more than happy to help, obviously, and I completely understand if you wish to postpone this work for a later time, if (when :) the user base expands and more people are available to do this important but unrewarding work.

**Comment by jeromelecoq** [comment]
Thanks for your thoughts!

I am still considering deploying this as a cloud service if there is enough demand. The code was somehow organized toward that goal. if you look at the notebook examples, most of the code is creating json files that are submitted to a few functions.  

For inference, I think ModelServer objects are often used in production. Since our models are standard TensorFlow networks, they can leverage this framework easily and essentially inference could be done bypassing most of DeepInterpolation codebase (perhaps except for the very little pre-processing code). My idea was to deploy the ModelServer onto an AWS fargate object or similar ECS service. 

I am currently working on a new figure for transfer training which should illuminate the balance between inference and training. I think this is the most important bit of info for deployment to the whole field: How much fine-tuning on each dataset is required when changing the recordings conditions further away from the Allen ?

**Comment by jeromelecoq** [comment]
Just an update on this thread. I am moving forward with implementing docker deployment on AWS sagemaker. Would love it if you want to test!

**Comment by jeromelecoq** [comment]
This was fixed! We completely revamped and introduced a new CLI. 
please go to the example folder for how to use OR type down : 
python -m deepinterpolation.cli.training --help

**Comment by jeromelecoq** [comment]
You should get something like this : 
usage: training.py [-h] [--input_json INPUT_JSON] [--output_json OUTPUT_JSON] [--log_level LOG_LEVEL] [--run_uid RUN_UID]
                   [--output_full_args OUTPUT_FULL_ARGS] [--training_params.output_dir TRAINING_PARAMS.OUTPUT_DIR]
                   [--training_params.name TRAINING_PARAMS.NAME]
                   [--training_params.steps_per_epoch TRAINING_PARAMS.STEPS_PER_EPOCH]
                   [--training_params.nb_times_through_data TRAINING_PARAMS.NB_TIMES_THROUGH_DATA]
                   [--training_params.learning_rate TRAINING_PARAMS.LEARNING_RATE]
                   [--training_params.loss TRAINING_PARAMS.LOSS] [--training_params.model_string TRAINING_PARAMS.MODEL_STRING]
                   [--training_params.caching_validation TRAINING_PARAMS.CACHING_VALIDATION]
                   [--training_params.multi_gpus TRAINING_PARAMS.MULTI_GPUS]
                   [--training_params.apply_learning_decay TRAINING_PARAMS.APPLY_LEARNING_DECAY]
                   [--training_params.epochs_drop TRAINING_PARAMS.EPOCHS_DROP]
                   [--training_params.period_save TRAINING_PARAMS.PERIOD_SAVE]
                   [--generator_params.data_path GENERATOR_PARAMS.DATA_PATH] [--generator_params.name GENERATOR_PARAMS.NAME]
                   [--generator_params.pre_frame GENERATOR_PARAMS.PRE_FRAME]
                   [--generator_params.post_frame GENERATOR_PARAMS.POST_FRAME]
                   [--generator_params.pre_post_omission GENERATOR_PARAMS.PRE_POST_OMISSION]
                   [--generator_params.batch_size GENERATOR_PARAMS.BATCH_SIZE]
                   [--generator_params.start_frame GENERATOR_PARAMS.START_FRAME]
                   [--generator_params.end_frame GENERATOR_PARAMS.END_FRAME]
                   [--generator_params.randomize GENERATOR_PARAMS.RANDOMIZE]
                   [--generator_params.total_samples GENERATOR_PARAMS.TOTAL_SAMPLES]
                   [--test_generator_params.data_path TEST_GENERATOR_PARAMS.DATA_PATH]
                   [--test_generator_params.name TEST_GENERATOR_PARAMS.NAME]
                   [--test_generator_params.pre_frame TEST_GENERATOR_PARAMS.PRE_FRAME]
                   [--test_generator_params.post_frame TEST_GENERATOR_PARAMS.POST_FRAME]
                   [--test_generator_params.pre_post_omission TEST_GENERATOR_PARAMS.PRE_POST_OMISSION]
                   [--test_generator_params.batch_size TEST_GENERATOR_PARAMS.BATCH_SIZE]
                   [--test_generator_params.start_frame TEST_GENERATOR_PARAMS.START_FRAME]
                   [--test_generator_params.end_frame TEST_GENERATOR_PARAMS.END_FRAME]
                   [--test_generator_params.randomize TEST_GENERATOR_PARAMS.RANDOMIZE]
                   [--test_generator_params.total_samples TEST_GENERATOR_PARAMS.TOTAL_SAMPLES]
                   [--network_params.name NETWORK_PARAMS.NAME]

optional arguments:
  -h, --help            show this help message and exit

TrainingInputSchema:
  --input_json INPUT_JSON
                        file path of input json file
  --output_json OUTPUT_JSON
                        file path to output json file
  --log_level LOG_LEVEL
                        set log level (default=INFO)
  --run_uid RUN_UID     unique identifier (default=2021_11_16_19_12)
  --output_full_args OUTPUT_FULL_ARGS
                        whether to output the full set of args to a json. this will show the args sent to the underlying
                        classes including defaults. (default=False)

training_params:
  --training_params.output_dir TRAINING_PARAMS.OUTPUT_DIR
                        A folder where the training outputs will get written. (REQUIRED)
  --training_params.name TRAINING_PARAMS.NAME
                        Training class to use. All available classes are visible in the trainor_collection.py file as part of
                        the deepinterpolation package. More classes can be added to this file to modify details of the
                        training behavior. (default=core_trainer) (valid options are ['core_trainer', 'transfer_trainer'])
  --training_params.steps_per_epoch TRAINING_PARAMS.STEPS_PER_EPOCH
                        Number of batches per epoch. Here, epochs are not defined in relation to the total number of batches
                        in the dataset as our datasets can be very large and it is beneficial to save models and evaluate
                        validation loss during training. After each epoch a validation loss is computed and a checkpoint model
                        is potentialy saved (see period_save). (default=100)
  --training_params.nb_times_through_data TRAINING_PARAMS.NB_TIMES_THROUGH_DATA
                        Setting this to more than 1 will make the training use individual batches multiple times during
                        training, thereby increasing your training samples. Larger repetition of the same batches could cause
                        noise overfitting (default=1)
  --training_params.learning_rate TRAINING_PARAMS.LEARNING_RATE
                        base learning rate used by the optimizer (default=0.0001)
  --training_params.loss TRAINING_PARAMS.LOSS
                        loss function used for training and validation. Loss functions recognized by tensorflow are recognized
                        : https://www.tensorflow.org/api_docs/python/tf/keras/losses. Additional losses can be added to the
                        loss_collection.py file. (default=mean_squared_error)
  --training_params.model_string TRAINING_PARAMS.MODEL_STRING
                        Text string used to save the final model file and all intermediary checkpoint models. Filename is
                        constructed from other fields if empty, using <network_name>_<loss>. (default=)
  --training_params.caching_validation TRAINING_PARAMS.CACHING_VALIDATION
                        Whether to cache the validation data in memory for training. On some systems, this could accelerate
                        training as it reduces the need for IO. On some system, the additional memory requirement could cause
                        memory issues. (default=False)
  --training_params.multi_gpus TRAINING_PARAMS.MULTI_GPUS
                        Set to True to use multi-gpus code when multi-gpus are available and set up on the machine and
                        environment. Single GPU or CPU code is used if set to False. (default=False)
  --training_params.apply_learning_decay TRAINING_PARAMS.APPLY_LEARNING_DECAY
                        whether to use a learning scheduler during training. If set to True, the learning rate will be halved
                        every <epochs_drop> (default=False)
  --training_params.epochs_drop TRAINING_PARAMS.EPOCHS_DROP
                        Number of epochs. Used when apply_learning_decay is set to True. Will halve the learning rate every
                        epoch_drop. One epoch is defined using steps_per_epoch. (default=5)
  --training_params.period_save TRAINING_PARAMS.PERIOD_SAVE
                        Period in number of epochs to periodically save model checkpoints. (default=5)

generator_params:
  --generator_params.data_path GENERATOR_PARAMS.DATA_PATH
                        Path to the file containing data used by the generator. Usually this will be a full filepath. In some
                        cases, this can point to a folder (with MultiContinuousTifGenerator) (REQUIRED)
  --generator_params.name GENERATOR_PARAMS.NAME
                        The data generator will control how data is read from individual data files. Specify a data generator
                        available in generator_collection.py, depending on your file data format. (default=SingleTifGenerator)
                        (valid options are ['EphysGenerator', 'MovieJSONGenerator', 'MultiContinuousTifGenerator',
                        'OphysGenerator', 'SingleTifGenerator'])
  --generator_params.pre_frame GENERATOR_PARAMS.PRE_FRAME
                        Number of frames fed to the DeepInterpolation model before a center frame for interpolation. Omitted
                        frames will not be used to fetch pre_frames. All pre_frame frame(s) will be fetched before
                        pre_omission frame(s). (default=30)
  --generator_params.post_frame GENERATOR_PARAMS.POST_FRAME
                        Number of frames fed to the DeepInterpolation model after a center frame for interpolation. Omitted
                        frames will not be used to fetch post_frames. All post_frame frame(s) will be fetch after
                        post_omission frame(s). (default=30)
  --generator_params.pre_post_omission GENERATOR_PARAMS.PRE_POST_OMISSION
                        Number of frames omitted before and after a center frame for DeepInterpolation. Omission will be done
                        on both sides of the center frame, ie. twice pre_post_omission are omitted. Omitted frames will not be
                        used to fetch pre_frames and post_frames. (default=0)
  --generator_params.batch_size GENERATOR_PARAMS.BATCH_SIZE
                        Batch size provided to the DeepInterpolation model by the generator. (default=5)
  --generator_params.start_frame GENERATOR_PARAMS.START_FRAME
                        First frame used by the generator. First frame is 0. (default=0)
  --generator_params.end_frame GENERATOR_PARAMS.END_FRAME
                        Last frame used by the generator. -1 defaults to the last available frame. Negative values smaller
                        than -1 increasingly truncates the end of the dataset. 0 is not permitted. Note: if end_frame = 2000,
                        frame 2000 will be included if there is sufficient post_frame frames available after frame 2000.
                        (default=-1)
  --generator_params.randomize GENERATOR_PARAMS.RANDOMIZE
                        Whether to shuffle all selected frames in the generator. It is recommended to set to 'True' for
                        training and 'False' for inference. (default=True)
  --generator_params.total_samples GENERATOR_PARAMS.TOTAL_SAMPLES
                        Total number of frames used between start_frame and end_frame. -1 defaults to all available samples
                        between start_frame and end_frame. If total_samples is larger than the number of available frames, it
                        will automatically be reduced to the maximal number. (default=-1)

test_generator_params:
  --test_generator_params.data_path TEST_GENERATOR_PARAMS.DATA_PATH
                        Path to the file containing data used by the generator. Usually this will be a full filepath. In some
                        cases, this can point to a folder (with MultiContinuousTifGenerator) (REQUIRED)
  --test_generator_params.name TEST_GENERATOR_PARAMS.NAME
                        The data generator will control how data is read from individual data files. Specify a data generator
                        available in generator_collection.py, depending on your file data format. (default=SingleTifGenerator)
                        (valid options are ['EphysGenerator', 'MovieJSONGenerator', 'MultiContinuousTifGenerator',
                        'OphysGenerator', 'SingleTifGenerator'])
  --test_generator_params.pre_frame TEST_GENERATOR_PARAMS.PRE_FRAME
                        Number of frames fed to the DeepInterpolation model before a center frame for interpolation. Omitted
                        frames will not be used to fetch pre_frames. All pre_frame frame(s) will be fetched before
                        pre_omission frame(s). (default=30)
  --test_generator_params.post_frame TEST_GENERATOR_PARAMS.POST_FRAME
                        Number of frames fed to the DeepInterpolation model after a center frame for interpolation. Omitted
                        frames will not be used to fetch post_frames. All post_frame frame(s) will be fetch after
                        post_omission frame(s). (default=30)
  --test_generator_params.pre_post_omission TEST_GENERATOR_PARAMS.PRE_POST_OMISSION
                        Number of frames omitted before and after a center frame for DeepInterpolation. Omission will be done
                        on both sides of the center frame, ie. twice pre_post_omission are omitted. Omitted frames will not be
                        used to fetch pre_frames and post_frames. (default=0)
  --test_generator_params.batch_size TEST_GENERATOR_PARAMS.BATCH_SIZE
                        Batch size provided to the DeepInterpolation model by the generator. (default=5)
  --test_generator_params.start_frame TEST_GENERATOR_PARAMS.START_FRAME
                        First frame used by the generator. First frame is 0. (default=0)
  --test_generator_params.end_frame TEST_GENERATOR_PARAMS.END_FRAME
                        Last frame used by the generator. -1 defaults to the last available frame. Negative values smaller
                        than -1 increasingly truncates the end of the dataset. 0 is not permitted. Note: if end_frame = 2000,
                        frame 2000 will be included if there is sufficient post_frame frames available after frame 2000.
                        (default=-1)
  --test_generator_params.randomize TEST_GENERATOR_PARAMS.RANDOMIZE
                        Whether to shuffle all selected frames in the generator. It is recommended to set to 'True' for
                        training and 'False' for inference. (default=True)
  --test_generator_params.total_samples TEST_GENERATOR_PARAMS.TOTAL_SAMPLES
                        Total number of frames used between start_frame and end_frame. -1 defaults to all available samples
                        between start_frame and end_frame. If total_samples is larger than the number of available frames, it
                        will automatically be reduced to the maximal number. (default=-1)

network_params:
  --network_params.name NETWORK_PARAMS.NAME
                        callback of the neuronal network architecture to build. All available architectures are visible in the
                        network_collection.py file as part of the deepinterpolation package. More architectures callbacks can
                        be added to this file if necessary. (default=unet_single_1024) (REQUIRED) (valid options are
                        ['autoencoder_single_256', 'dense_thick_units', 'fmri_flexible_architecture', 'fmri_unet_denoiser',
                        'fmri_volume_deeper_denoiser', 'fmri_volume_denoiser', 'fmri_volume_dense_denoiser',
                        'fmri_volume_optimized_denoiser', 'padding_unet_single_1024', 'segmentation_net', 'unet_1024_search',
                        'unet_double_1024', 'unet_single_1024', 'unet_single_1p_1024', 'unet_single_256',
                        'unet_single_ephys_1024'])

**Comment by jeromelecoq** [comment]
I am going to close this issue
